src/container/srv_container.c

```diff
@@ -960,6 +971,265 @@ cont_destroy(struct rdb_tx *tx, struct ds_pool_hdl *pool_hdl,
+ */
+int
+ds_cont_update_agg_eph(uuid_t pool_uuid, uuid_t cont_uuid,
+		       d_rank_t rank, daos_epoch_t eph)
+{
+	struct cont_svc		*svc;
+	struct cont_ec_agg	*ec_agg;
+	int			rc;
+	int			i;
+
+	rc = cont_svc_lookup_leader(pool_uuid, 0 /* id */, &svc,
+				    NULL /* hint */);
+	if (rc != 0)
+		return rc;
+
+	ec_agg = cont_ec_agg_lookup(svc, cont_uuid);
+	if (ec_agg == NULL) {
+		rc = cont_ec_agg_alloc(svc, cont_uuid, &ec_agg);
+		if (rc)
+			D_GOTO(out_put, rc);
+	}
+
+	for (i = 0; i < ec_agg->ea_servers_num; i++) {
+		if (ec_agg->ea_server_ephs[i].rank == rank){
+			if (ec_agg->ea_server_ephs[i].eph < eph)
+				ec_agg->ea_server_ephs[i].eph = eph;
+			break;
+		}
+	}
+
+	D_DEBUG(DB_MD, DF_CONT" update agg eph rank %u eph "DF_U64"\n",
+		DP_CONT(pool_uuid, cont_uuid), rank, eph);
+	if (i == ec_agg->ea_servers_num)
+		D_WARN("rank %u eph "DF_U64" does not exist for "DF_CONT"\n",
+			rank, eph, DP_CONT(pool_uuid, cont_uuid));
+out_put:
+	cont_svc_put_leader(svc);
+	return 0;
+}
+
+struct refresh_vos_agg_eph_arg{
+	uuid_t	pool_uuid;
+	uuid_t  cont_uuid;
+	daos_epoch_t min_eph;
+};
+
+int
+cont_refresh_vos_agg_eph_one(void *data)
+{
+	struct refresh_vos_agg_eph_arg *arg = data;
+	struct ds_cont_child	*cont_child;
+	int			rc;
+
+	rc = ds_cont_child_lookup(arg->pool_uuid, arg->cont_uuid, &cont_child);
+	if (rc)
+		return rc;
+
+	D_DEBUG(DB_MD, DF_CONT": update aggregation max eph "DF_U64"\n",
+		DP_CONT(arg->pool_uuid, arg->cont_uuid), arg->min_eph);
+	cont_child->sc_aggregation_max = arg->min_eph;
+	ds_cont_child_put(cont_child);
+	return rc;
+}
+
+int
+ds_cont_refresh_agg_eph(uuid_t pool_uuid, uuid_t cont_uuid, daos_epoch_t eph)
+{
+	struct refresh_vos_agg_eph_arg	arg;
+	int				rc;
+
+	uuid_copy(arg.pool_uuid, pool_uuid);
+	uuid_copy(arg.cont_uuid, cont_uuid);
+	arg.min_eph = eph;
+
+	rc = dss_task_collective(cont_refresh_vos_agg_eph_one, &arg, 0,
+				 DSS_ULT_IO);
+	return rc;
+}
+
+#define EC_AGG_EPH_INTV	 (10ULL * NSEC_PER_SEC)	/* seconds interval to check*/
+static void
+cont_agg_eph_leader_ult(void *arg)
+{
+	struct dss_sleep_ult	*sleep_ult;
+	struct cont_svc		*svc = arg;
+	int			rc = 0;
+
+	sleep_ult = dss_sleep_ult_create();
+	if (sleep_ult == NULL) {
+		D_ERROR(DF_UUID " sleep ult creation failed.\n",
+			DP_UUID(svc->cs_pool->sp_uuid));
+		return;
+	}
+
+	while(1) {
+		struct cont_ec_agg *ec_agg;
+
+		d_list_for_each_entry(ec_agg, &svc->cs_ec_agg_list,
+				      ea_list) {
+			daos_epoch_t min_eph = DAOS_EPOCH_MAX;
+			int	     i;
+
+			if (svc->cs_pool->sp_stopping)
+				break;
+
+			for (i = 0; i < ec_agg->ea_servers_num; i++) {
+				if (ec_agg->ea_server_ephs[i].eph < min_eph)
+					min_eph = ec_agg->ea_server_ephs[i].eph;
+			}
+
+			if (min_eph <= ec_agg->ea_current_eph)
+				continue;
+
+			D_DEBUG(DB_MD, DF_CONT" sync "DF_U64"\n",
+				DP_CONT(svc->cs_pool_uuid, ec_agg->ea_cont_uuid),
+				min_eph);
+			rc = cont_iv_ec_agg_eph_refresh(
+						svc->cs_pool->sp_iv_ns,
+						ec_agg->ea_cont_uuid, min_eph);
```

Comment:
every server will call cont_iv_ec_agg_eph_update asynchronously, here how to ensure it call cont_iv_ec_agg_eph_refresh() only after received all servers' report? (or you mean need not wait all servers' report done?)

Reply:
oh, no, it does not to be synchronize to wait all servers to report done, since the epoch on each server is increasing, so the leader just keep walking through the epoch for each ranks and get the minimum rank, then sync it all servers.

Reply:
ok, the "if (min_eph == ec_agg->ea_current_eph) continue" actually is the wait.
But need to skip failed node to avoid never can move ahead as above discuss.